### ISSUE TITLE
Can these features assist Downstream Training 

### ISSUE THREAD

**Issue by Mukil07** [WalterSimoncini/fungivision#3]
Suppose the fungi features are computed and passed to the rest of the downstream network train on their losses, keeping with the fungi feature extraction network exactly the same. Will this improve the downstream performance ?

**Comment by WalterSimoncini** [comment]
Hey @Mukil07! So, if I understood correctly you'd want to:

1. Extract FUNGI features from a pre-trained backbone
2. Use them as the input for a second network to solve a downstream task

If that's the case yup, they should work nicely, as long as you make sure to use an appropriate loss to generate the FUNGI features. For example, if you just need global information about an image (as in e.g. classification) you should be good to go with the code/losses in this library. If you need fine-grained information (e.g. for semantic segmentation) you should use a dense objective to build FUNGIs, e.g. in the code linked below we use a SimCLR-style loss over the patch tokens.

https://github.com/WalterSimoncini/no-train-all-gain/blob/main/fungi-hummingbird/eval_gradients.py#L73

This is currently not implemented in the library, but it shouldn't be too hard to include it.

**Comment by Mukil07** [comment]
Thank you that explains my questions. Could you please answer the following questions: 

1. can we extract fungi features for multiscale image features, is it possible to extract them without flattening the spatial dimensions, which should give fungi feature maps? 

2. If I get it right, the reason we use PCA is because the goal is to not train. So can we just straight up use Linear linear and learn the conversion from higher dim space (after concatenation) to original embedding dim ?

**Comment by WalterSimoncini** [comment]
Awesome, and you're welcome! As for your follow up questions:

1. Most likely. We flatten/vectorize the gradients (and downsample them with a random projection) since by default we use the gradients with respect to a linear layer's weight (which, afaik, have no particular meaning as a map/grid). For multiscale image features you should probably add a torch hook to capture the gradients with respect to the layer's output (or input) and use those as FUNGI features (which will have the same dimensionality as the multi-scale features). Just make sure to use all the feature maps in the loss computation, otherwise you may end up with some of them having zero gradients.
2. Yes, the main reason we use PCA is to make sure FUNGIs have the same dimensionality as the embeddings (mainly for storage reasons). This also provides a small performance boost, as you can see in Table 21 [in the paper](https://arxiv.org/pdf/2407.10964). You can also simpy use a random projection here, but it will perform slightly worse.

Let me know if this is clear or if you have any further questions :)

**Comment by Mukil07** [comment]
Thank you so much !! for those comments. I'll close this issue
